packages/perf/src/index.ts

```diff
@@ -0,0 +1,166 @@
+
+    let bytesToSendBack: bigint | null = null
+
+    for await (const buf of stream.source) {
+      if (bytesToSendBack === null) {
+        bytesToSendBack = BigInt(buf.getBigUint64(0, false))
+      }
+      // Ingest all the bufs and wait for the read side to close
+    }
+
+    const uint8Buf = new Uint8Array(this.databuf)
+
+    if (bytesToSendBack === null) {
+      throw new Error('bytesToSendBack was null')
+    }
+
+    await stream.sink(async function * () {
+      while (bytesToSendBack > 0n) {
+        let toSend: bigint = writeBlockSize
+        if (toSend > bytesToSendBack) {
+          toSend = bytesToSendBack
+        }
+        bytesToSendBack = bytesToSendBack - toSend
+        yield uint8Buf.subarray(0, Number(toSend))
+      }
+    }())
+  }
+
+  async perf (peer: PeerId, sendBytes: bigint, recvBytes: bigint, options: AbortOptions = {}): Promise<void> {
+    log('opening stream on protocol %s to %p', this.protocol, peer)
+
+    const uint8Buf = new Uint8Array(this.databuf)
+
+    const writeBlockSize = this.writeBlockSize
+
+    const connection = await this.components.connectionManager.openConnection(peer, options)
+
+    const signal = anySignal([AbortSignal.timeout(this.timeout), options?.signal])
+    const stream = await connection.newStream([this.protocol], {
+      signal
+    })
+
+    // Convert sendBytes to uint64 big endian buffer
+    const view = new DataView(this.databuf)
+    view.setBigInt64(0, recvBytes, false)
+
+    log('sending %i bytes to %p', sendBytes, peer)
+
+    await stream.sink((async function * () {
+      // Send the number of bytes to receive
+      yield uint8Buf.subarray(0, 8)
+      // Send the number of bytes to send
+      while (sendBytes > 0n) {
+        let toSend: bigint = writeBlockSize
+        if (toSend > sendBytes) {
+          toSend = sendBytes
+        }
+        sendBytes = sendBytes - toSend
+        yield uint8Buf.subarray(0, Number(toSend))
+      }
+    })())
+
+    // Read the received bytes
+    let actualRecvdBytes = BigInt(0)
+    for await (const buf of stream.source) {
+      actualRecvdBytes += BigInt(buf.length)
+    }
+
+    if (actualRecvdBytes !== recvBytes) {
+      throw new Error(`Expected to receive ${recvBytes} bytes, but received ${actualRecvdBytes}`)
+    }
+
+    log('performed %s to %p', this.protocol, peer)
+    stream.close()
+  }
+
+  // measureDownloadBandwidth returns the measured bandwidth in bits per second
+  async measureDownloadBandwidth (peer: PeerId, size: bigint): Promise<number> {
+    const now = Date.now()
+    await this.perf(peer, 0n, size)
+    return Number((8000n * size) / BigInt(Date.now() - now))
+  }
+
+  // measureUploadBandwidth returns the measured bandwidth in bit per second
+  async measureUploadBandwidth (peer: PeerId, size: bigint): Promise<number> {
+    const now = Date.now()
+    await this.perf(peer, size, 0n)
+    return Number((8000n * size) / BigInt(Date.now() - now))
+  }
+}
```

Comment:
Based on the `8000` I am assuming that `size`' unit is `kbyte`. For the sake of simplicity and easier understanding I suggest using [SI units](https://en.wikipedia.org/wiki/International_System_of_Units) everywhere. I.e. in this case I suggest using `byte` instead of `kbyte`.